### ISSUE TITLE
Magic strings are not to be uttered

### ISSUE THREAD

**Issue by dreeves** [beeminder/blog#371]
[placeholder while dreev thinks about this]

## Cognata

* #346 

_**Verbata**: don't accidentally invoke magic, dark magic,_

**Comment by narthur** [comment]
@dreeves Is this something like just adding a check to ensure BEGIN_MAGIC and END_MAGIC aren't ever in final post output?

**Comment by dreeves** [comment]
No, but I might need a whole blog post to explain it and it might just be dumb.

Now you've got me thinking about that blog post. Brainstorming:

You know how a magic spell is an incantation that, just by being incanted, makes something happen in the physical world? You say "avada kedavra" and someone dies -- that sort of thing. Well that's a literal real thing in programming and it's called a magic string. We use them in Beeminderland occasionally. The source document from which this blog post was generated uses them. That means that I absolutely cannot tell you what those magic strings are in this blog post. If I did it would literally break the blog and you would not be reading these words. Sounds dangerous, you say? Kind of! But it's pretty ok with a couple ground rules. One is that you only pick a magic string that can never possibly be uttered accidentally. Another is that you treat magic strings with some reverence and don't casually type them out if you don't mean to invoke them. Like how in the Harry Potter universe people refer to "the killing curse" rather than actually utter "avada kedavra" in casual conversation.

**Comment by dreeves** [comment]
Robert Perce:

the usage of "magic string" that i'm familiar with (a string constant that must simply be Known, for which the compiler can't help you) does not seem to line up with most of that description, so maybe i'm not sure about the thing you're talking about

DREEV: ah, fair. i'm talking about a subset of that where you count on the magic string's uniqueness. like having a special string to mark end-of-file and if you happened to use that string in a file then everything would break.

ah! i think you should work in the phrase "in-band signalling", then, and also mention famous failure cases:

* +++ATH0
* quotes in sql-injected user input
* EOT in user input
 
maybe something like "in many piece of media, wizards communicate spells by learning them backwards, or with the words shuffled around. That's how control characters must be talked about: I can tell you that 'the contents of my string are represented in ascii by bytes 98, 101, 101, 0, 109, 105, 111, 100', but I can't (in many languages) actually print that string out, because the NUL in the middle (another name for the-character-represented-in-ascii-by-byte-0) indicates end of string."

and a good opportunity to contrast with out of band signalling, e.g., "sql queries using ?" & the fact that in, e.g., C., if you get the value of EOF it's bigger than a byte so it's not possible to have an (ascii-encoded) file that attempts to _contain_ an EOF.

**Comment by dreeves** [comment]
![image](https://github.com/beeminder/blog/assets/9928/762a62b6-9baf-43e3-98eb-5823165945d7)

**Comment by narthur** [comment]
@dreeves I just took the DEV label off this issue, since it seems to be more of a blog post idea than anything at this point. Let me know if it should be re-added.

**Comment by bsoule** [comment]
<img width="580" height="580" alt="Image" src="https://github.com/user-attachments/assets/279176b3-48dc-4655-a2cb-f39f75009b07" />

**Comment by bsoule** [comment]
<img width="871" height="877" alt="Image" src="https://github.com/user-attachments/assets/8fc4201d-ab0c-4aa5-b260-e05ea0f4d0f2" />
